### PR TITLE
Fixed RouteDrawer

### PR DESCRIPTION
**Pull request by RoiEXLab**
Follow up of #1118 avoids large diff.
The RouteDrawer is working perfectly now! Finally...

---
This change is [Reviewable](https://reviewable.io/reviews/triplea-game/triplea/1132)

### CODE REVIEW

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 1]
```diff
@@ -40,14 +40,19 @@
...
    * Draws the route to the screen.
    */
   public void drawRoute(final Graphics2D graphics, final RouteDescription routeDescription, final MapPanel mapPanel,
       final MapData mapData, final String maxMovement) {
-    if (routeDescription == null) {
+    if (routeDescription == null || mapData == null || mapPanel == null) {
       return;
     }
+    if (routeOptimizer == null) {
+      routeOptimizer = new RouteOptimizer(mapData, mapPanel);
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: why cache this object and not recreate it every time?

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 2]
```diff
@@ -40,14 +40,19 @@
...
 
+  private RouteOptimizer routeOptimizer;
+
   /**
    * Draws the route to the screen.
    */
   public void drawRoute(final Graphics2D graphics, final RouteDescription routeDescription, final MapPanel mapPanel,
       final MapData mapData, final String maxMovement) {
-    if (routeDescription == null) {
+    if (routeDescription == null || mapData == null || mapPanel == null) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: suspect null checks, ideally we would just not call this method if these were null. Can we enforce that on the caller side?

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 3]
```diff
@@ -106,9 +101,11 @@ private void drawJoints(final Graphics2D graphics, final Point[] points, final i
     final int jointsize = 10;
     // If the points array is bigger than 1 the last joint should not be drawn (draw an arrow instead)
     final Point[] newPoints = points.length > 1 ? Arrays.copyOf(points, points.length - 1) : points;
-    for (final Point p : newPoints) {
-      graphics.fillOval((int) (((p.x - xOffset) - (jointsize / 2) / scale) * scale),
-          (int) (((p.y - yOffset) - (jointsize / 2) / scale) * scale), jointsize, jointsize);
+    for (Point[] joints : routeOptimizer.getAllPoints(newPoints)) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: Some commentary on what is happening will help us maintain this in the future.
Ideally this would be broken up, possibly extracted to a new method or class etc.. Minimum bar IMO is something to help a developer reading this to know what is going on.

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 2 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: We should, I remember adding those checks, when quick fixing the failing tests, which test if null values do throw exceptions...

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 4]
```diff
@@ -122,15 +119,14 @@ private void drawJoints(final Graphics2D graphics, final Point[] points, final i
...
     if (cursorImage != null) {
-      final Point wrappedEndPoint =
-          getPointOnMap(routeDescription.getEnd(), xOffset, yOffset, imageDimension, screenDimension);
-      graphics.drawImage(cursorImage,
-          (int) (((wrappedEndPoint.x - xOffset) - (cursorImage.getWidth(null) / 2)) * scale),
-          (int) (((wrappedEndPoint.y - yOffset) - (cursorImage.getHeight(null) / 2)) * scale), null);
+      for (Point[] endPoint : routeOptimizer.getAllPoints(routeOptimizer.getLastEndPoint())) {
+        graphics.drawImage(cursorImage,
+            (int) (((endPoint[0].x - xOffset) - (cursorImage.getWidth(null) / 2)) * scale),
+            (int) (((endPoint[0].y - yOffset) - (cursorImage.getHeight(null) / 2)) * scale), null);
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: What does this do? 
`cursorImage.getHeight(null)`

Can we avoid passing in null to a method and having it do some seemingly magic work? This could be a useful thing to move to a library/utility somewhere so when reading it we are not left wondering how that works (or what the `null` param does)

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 1 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: because neither the mapPanel nor the mapData change... there is no need for recreating...

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 5]
```diff
@@ -149,10 +145,13 @@ private void drawCustomCursor(final Graphics2D graphics, final RouteDescription
...
       final int yOffset, final double scale) {
-    drawLineWithTranslate(graphics, new Line2D.Float(start, end), xOffset,
-        yOffset, scale);
-    if (start.distance(end) > arrowLength) {
-      drawArrow(graphics, start, end, xOffset, yOffset, scale);
+    final Point[] points = routeOptimizer.getTranslatedRoute(start, end);
+    for (Point[] newPoints : routeOptimizer.getAllPoints(points)) {
+      drawLineWithTranslate(graphics, new Line2D.Float(newPoints[0], newPoints[1]), xOffset,
+          yOffset, scale);
+      if (newPoints[0].distance(newPoints[1]) > arrowLength) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: some commentary to explain the logic and reasoning would be helpful. The heavy usage of array indexes is a bit troubling, but meh..

File: src/games/strategy/triplea/ui/MapRouteDrawer.java [hunk 6]
```diff
@@ -203,58 +202,26 @@ private void drawLineWithTranslate(final Graphics2D graphics, final Line2D line,
...
-      final int yOffset,
-      final Dimension imageDimension, final Dimension screenDimension) {
+  protected Point[] getRoutePoints(final RouteDescription routeDescription, final MapData mapData) {
     final List<Territory> territories = routeDescription.getRoute().getAllTerritories();
     final int numTerritories = territories.size();
     final Point[] points = new Point[numTerritories];
     for (int i = 0; i < numTerritories; i++) {
-      points[i] =
-          getPointOnMap(mapData.getCenter(territories.get(i)), xOffset, yOffset, imageDimension, screenDimension);
+      points[i] = mapData.getCenter(territories.get(i));
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: :+1: 
nice.

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+  public RouteOptimizer(MapData mapData, MapPanel mapPanel) {
+    checkNotNull(mapData);
+    this.mapPanel = checkNotNull(mapPanel);
+    isInfiniteY = mapData.scrollWrapY();
+    isInfiniteX = mapData.scrollWrapX();
+  }
+
+  public Point[] getTranslatedRoute(Point... route) {
+    if (route == null) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: may want to check if the route is empty as well.
Moving this check to the caller IMO is ideal, simply do not call the method with no-op params.

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+      return route;
+    }
+    if (!isInfiniteX && !isInfiniteY || route.length == 0) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: Please put parens to do explicit grouping of the boolean conditions, ie: Does 'a & b || c' mean "(a & b) || c", or is it "a & (b || c)"

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+    for (Point point : route) {
+      if (previousPoint == null) {
+        previousPoint = point;
+        result.add(point);
+        continue;
+      }
+      previousPoint = normalizePoint(previousPoint);// TODO
+      double closestDistance = Double.MAX_VALUE;
+      Point closestPoint = null;
+      for (Point2D possibleClosestPoint : getPossiblePoints(point)) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: long method smell here, indentation is getting a bit deep.

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+          alternativePoints.get(6)[counter] = new Point(normalizedPoint.x + mapWidth, normalizedPoint.y - mapHeight);
+          alternativePoints.get(7)[counter] = new Point(normalizedPoint.x + mapWidth, normalizedPoint.y + mapHeight);
+        }
+        counter++;
+      }
+    }
+    return alternativePoints;
+  }
+
+  public List<Point[]> getAllPoints(Point... points) {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: javadoc on this public method would be useful.

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+    for (int i = 0; i < points.length; i++) {
+      normalizedPoints[i] = normalizePoint(points[i]);
+    }
+    allPoints.add(normalizedPoints);
+    return allPoints;
+  }
+
+  private int getAlternativePointArrayCount() {
+    if (isInfiniteX && isInfiniteY) {
+      return 8;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: magic numbers, please extract them to constants at the minimum to give them a name.

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 4 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: this is the only way to get the width/heigth of the BufferedImage object, don't ask me why...

File: test/games/strategy/triplea/ui/TestRoute.java [hunk 12]
```diff
@@ -73,20 +72,20 @@ public void testPointSplitting() {
...
-      assertArrayEquals(dummyPoints, spyRouteDrawer.getRoutePoints(dummyRouteDescription, dummyMapData, randomInt(),
-          randomInt(), new Dimension(), new Dimension()));
-      assertArrayEquals(dummyPoints, spyRouteDrawer.getRoutePoints(dummyRouteDescription, dummyMapData, 0, 0,
-          new Dimension(randomInt(1000, 100), randomInt(1000, 100)), new Dimension()));
-      final int randX = randomInt(1000, 0);
-      assertEquals(new Point(randX, 0),
-          MapRouteDrawer.getPointOnMap(new Point(randX, 0), randX, 0, new Dimension(randX, 0), new Dimension()));
-    }
-  }
+//  @Test
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/ui/TestRoute.java]: what's the context  for commenting out the test?

File: test/games/strategy/triplea/ui/TestRoute.java [hunk 13]
```diff
@@ -97,6 +96,7 @@ private static int randomInt(int max, int min) {
     return (int) ((Math.random() * (max - min)) + min);
   }
 
+  @SuppressWarnings("unused")
```
Comment (DanVanAtta) [inline test/games/strategy/triplea/ui/TestRoute.java]: remove this instead of suppressing the warning?

File: src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java (hunk 8 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/mapdata/RouteOptimizer.java]: Oh I see the mistake... This is unintended

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 4 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: create a private static function to hide that detail? At least then the confusion will be buried in a deeper layer where we can comment on it explicitly.

File: test/games/strategy/triplea/ui/TestRoute.java (hunk 13 above)
Comment (RoiEXLab) [inline test/games/strategy/triplea/ui/TestRoute.java]: I have to rework the tests anyway. Removing this once the tests are working again...

File: test/games/strategy/triplea/ui/TestRoute.java (hunk 12 above)
Comment (RoiEXLab) [inline test/games/strategy/triplea/ui/TestRoute.java]: This is no longer working since the method does no longer exist... Will remove on test fixing

**Comment by RoiEXLab** [comment]
Thanks @DanVanAtta I will continue working on this tomorrow, when I have slept a bit. Although it's fun coding at about 0:00h it's not very efficient...

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 1 above)
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: Objects are super cheap in Java8, and routeOptimizer is basically only holding references, the cost of recreating it is negligible.  The caching is simply not necessary, and adds state to the class, which complicates things more than we are helping.

Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: But whe would still need the field to avoid passing mapData objects all over the place.

File: src/games/strategy/triplea/ui/MapRouteDrawer.java (hunk 4 above)
Comment (RoiEXLab) [inline src/games/strategy/triplea/ui/MapRouteDrawer.java]: Better Idea... BufferedImage provides a getWidth/getHeight method casting fixes this

**Comment by RoiEXLab** [comment]
Oh I forgot... The obligatory updated Message
@DanVanAtta

**Comment by DanVanAtta** [comment]
I'll review in detail later tonight. If things work out well, this is the last thing to merge before we have a real release candidate.

File: src/games/strategy/triplea/ui/RouteOptimizer.java [hunk 14]
```diff
@@ -18,6 +18,10 @@
 
   private final MapPanel mapPanel;
 
+  private static final int maxAdditionalScreens = 8;
+  private static final int commonAdditionalScreens = 2;
+  private static final int minAdditionalScreens = 0;
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/RouteOptimizer.java]: formatting nit: constants / immutable class data should be in caps

File: src/games/strategy/triplea/ui/RouteOptimizer.java [hunk 15]
```diff
@@ -0,0 +1,250 @@
...
+import java.awt.Point;
+import java.awt.geom.Line2D;
+import java.awt.geom.Point2D;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import games.strategy.triplea.ui.mapdata.MapData;
+
+public class RouteOptimizer {
```
Comment (DanVanAtta) [inline src/games/strategy/triplea/ui/RouteOptimizer.java]: Looks like a good decomposition :+1:   Noting that this class is much easier to test than MapRouteDrawer, and also makes MapRouteDrawer since we can mock the logic here and return hardcoded values so we do not have to worry about the calculations when testing MapRouteDrawer.
